File: LC/Medium/380.py
class RandomizedSet:

    def __init__(self):
        self.data = []
        self.dict = {}

    def insert(self, val: int) -> bool:
        if val in self.dict:
            return False
        
        self.dict[val] = len(self.data)
        self.data.append(val)
        
        return True

    def remove(self, val: int) -> bool:
        if val not in self.data:
            return False
        
        index = self.dict[val]
        last_element = self.data[-1]
        
        self.data[index] = last_element
        self.dict[last_element] = index
        
        self.data.pop()
        del(self.dict[val])
        
        return True
        
    def getRandom(self) -> int:
        return random.choice(self.data)
        


# Your RandomizedSet object will be instantiated and called as such:
# obj = RandomizedSet()
# param_1 = obj.insert(val)
# param_2 = obj.remove(val)
# param_3 = obj.getRandom()

File: LC/C++/Medium/380.cpp
class RandomizedSet {
    std::vector<int> data_;
    std::unordered_map<int, int> index_;

public:
    RandomizedSet() {}
    
    bool insert(int val) {
        if (index_.contains(val)) return false;

        index_[val] = data_.size();
        data_.push_back(val);

        return true;
    }
    
    bool remove(int val) {
        if (!index_.contains(val)) return false;

        int last = data_.back();
        int index = index_[val];

        data_[index] = last;
        index_[last] = index;

        data_.pop_back();
        index_.erase(val);

        return true;
    }
    
    int getRandom() {
        return data_[rand() % data_.size()];
    }
};

/**
 * Your RandomizedSet object will be instantiated and called as such:
 * RandomizedSet* obj = new RandomizedSet();
 * bool param_1 = obj->insert(val);
 * bool param_2 = obj->remove(val);
 * int param_3 = obj->getRandom();
 */